File: problem-solving/_lc-3-medium-longest-substring-without-repeating-characters/1-cubic.go
/*
https://leetcode.com/problems/longest-substring-without-repeating-characters

Complexity
----------
Time: O(n^3)
Space: O(n)

NOTE: This solution is not accepted, gives TLE
*/

type void struct{}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func allUniq(start, end int, s string) bool {
    set := make(map[byte]void)
    var null void
    for i := start; i < end + 1; i++ {
        if _, ok := set[s[i]]; ok {
            return false
        }
        set[s[i]] = null
    }
    return true
}

func lengthOfLongestSubstring(s string) int {
    ans := 0
    n := len(s)
    for i := 0; i < n; i++ {
        for j := i; j < n; j++ {
            if allUniq(i, j, s) {
                ans = max(ans, j - i + 1)
            }
        }
    }
    return ans
}

File: problem-solving/lc-3-medium-longest-substring-without-repeating-characters/1-cubic.py
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters

Complexity
----------
Time: O(n^3)
Space: O(n)

NOTE: This solution is not accepted, gives TLE
"""

def allUniq(start, end, s):
    seen = set()
    for i in range(start, end + 1):
        if s[i] in seen:
            return False
        seen.add(s[i])
    return True

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        mx = 0
        for i in range(n):
            for j in range(i, n):
                if allUniq(i, j, s):
                    mx = max(mx, j - i + 1)
        return mx
